usual/tls/tls.c

```diff
@@ -742,4 +742,84 @@ tls_close(struct tls *ctx)
 	return (rv);
 }
 
+static bool
+tls_config_keypair_equal(struct tls_keypair *left_tls_keypair, struct tls_keypair *right_tls_keypair)
+{
+  if (!strings_equal(left_tls_keypair->cert_file, right_tls_keypair->cert_file))
+    return false;
+  if (left_tls_keypair->cert_mem && right_tls_keypair->cert_mem &&
+      !(memcmp(left_tls_keypair->cert_mem, right_tls_keypair->cert_mem, left_tls_keypair->cert_len) == 0))
+    return false;
+  if (!strings_equal(left_tls_keypair->key_file, right_tls_keypair->key_file))
+    return false;
+  if (left_tls_keypair->key_mem && right_tls_keypair->key_mem &&
+      !(memcmp(left_tls_keypair->key_mem, right_tls_keypair->key_mem, left_tls_keypair->key_len) != 0))
+    return false;
+  return true;
+}
+
+static bool
+tls_config_keypairs_equal(struct tls_keypair *tls_keypair_left, struct tls_keypair *tls_keypair_right)
+{
+	bool keypair_unchanged;
+	struct tls_keypair *tls_kp_lt, *next_tls_kp_lt, *tls_kp_rt, *next_tls_kp_rt;
+
+	keypair_unchanged = tls_config_keypair_equal(tls_keypair_left, tls_keypair_right);
+
+	if (keypair_unchanged == false) {
+		return false;
+	}
+
+	for (tls_kp_lt = tls_keypair_left; tls_kp_lt != NULL; tls_kp_lt = next_tls_kp_lt) {
+		next_tls_kp_lt = tls_kp_lt->next;
+		for (tls_kp_rt = tls_keypair_right; tls_kp_rt != NULL; tls_kp_rt = next_tls_kp_rt) {
+			next_tls_kp_rt = tls_kp_rt->next;
+
+			keypair_unchanged = tls_config_keypair_equal(tls_kp_lt, tls_kp_rt);
+		}
+	}
+
+	return keypair_unchanged;
+}
+
+bool
+tls_configs_equal(struct tls_config *tls_config_left, struct tls_config *tls_config_right)
+{
+  if (!(strings_equal(tls_config_left->ca_file, tls_config_right->ca_file)))
+    return false;
+  if (!(strings_equal(tls_config_left->ca_path, tls_config_right->ca_path)))
+    return false;
+  if (tls_config_left->ca_mem &&  tls_config_right->ca_mem &&
+      !(memcmp(tls_config_left->ca_mem, tls_config_right->ca_mem, tls_config_left->ca_len) == 0))
```

Comment:
Let's move this logic to a `tls_mems_equal` function, so it can be reused for each of the 4 different `xyz_mem` fields.

Reply:
Also, to avoid out-of-bound reads. Before the `memcmp` we should check that the `ca_len` fields are equal.

Reply:
There are 2 occurrences and I don't know if it is worth. I suggest the following logic for both (`ca_mem` and `ocsp_mem`).

```
	if (tc1->ca_len != tc2->ca_len)
		return false;

	if (tc1->ca_mem == NULL && tc2->ca_mem == NULL)
		;
	else if (tc1->ca_mem == NULL || tc2->ca_mem == NULL)
		return false;
	else if (memcmp(tc1->ca_mem, tc2->ca_mem, tc1->ca_len) != 0)
		return false;
```

Reply:
There's also two occurences in `tls_config_keypair_equal`. So we'd need 4 copies, which makes it worth a function imho.

Reply:
I like it

Reply:
I will also check the len field equality

Reply:
ffe0447 and 768681